### PR TITLE
- Removed existing logger scopes as we want to minimize the number of sc...

### PR DESCRIPTION
**Pull request by kichalla**
...opes being created.

- Cleaned up tests related to removal of scopes.

@rynowak

### CODE REVIEW

**Comment by rynowak** [comment]
Removing the scopes is a good start. I think we could even live without the logging in router-middleware and route-collection as well -- @loudej thoughts?

We want logging when there's an important event that happens - in our code that happens inside `TemplateRoute` pretty much exclusively. The middleware and the collection really just provide aggregation, and that's the kind of thing we want to cut.

**Comment by rynowak** [comment]
I'd also really like to see diversification in the kinds of messages logged by `TemplateRoute` `CreateRouteAsyncValues` right now produces a one-size-fits-all message. Can you come up with a list of possible scenarios for `TemplateRoute` and we can think about which data goes with each one?

**Comment by kichalla** [comment]
Some examples:
1. 	A request resulted in desired response
2. 	A request resulted in 404 response
	- Possibilities:
		- Routes did not match due to template mismatch
		- Routes did not match due to constraints failure.
		- Route matched but application layer explicitly returned a 404 (ex: product not found)
		- Route matched but no controller/action present
3. 	A request resulted in undesired response.
	- Possiblities:
		- Route matched but resulted in an unexpected controller/action getting executed.

File: src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs [hunk 1]
```diff
@@ -29,30 +30,21 @@ public static class RouteConstraintMatcher
...
-                        logger.WriteValues(new RouteConstraintMatcherMatchValues()
+                        string currentRouteValues = string.Empty;
+                        if (routeValues != null)
                         {
-                            ConstraintKey = kvp.Key,
-                            Constraint = kvp.Value,
-                            Matched = false
-                        });
+                           currentRouteValues  = string.Join(", ", 
+                               routeValues.Select(rv => $"Key={rv.Key};Value={rv.Value}"));
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs]: We don't want to 'encode' values for logging. Just log the value that's relevant to the constraint.

File: src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs [hunk 2]
```diff
@@ -29,30 +30,21 @@ public static class RouteConstraintMatcher
                 var constraint = kvp.Value;
                 if (!constraint.Match(httpContext, route, kvp.Key, routeValues, routeDirection))
                 {
-                    if (routeDirection.Equals(RouteDirection.IncomingRequest)
-                        && logger.IsEnabled(LogLevel.Verbose))
+                    if (routeDirection.Equals(RouteDirection.IncomingRequest))
                     {
-                        logger.WriteValues(new RouteConstraintMatcherMatchValues()
+                        string currentRouteValues = string.Empty;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs]: var

File: src/Microsoft.AspNet.Routing/RouterMiddleware.cs [hunk 3]
```diff
@@ -30,22 +30,16 @@ public class RouterMiddleware
...
-                await _router.RouteAsync(context);
+            await _router.RouteAsync(context);
 
-                if (_logger.IsEnabled(LogLevel.Verbose))
-                {
-                    _logger.WriteValues(new RouterMiddlewareInvokeValues() { Handled = context.IsHandled });
-                }
+            if (!context.IsHandled)
+            {
+                _logger.LogWarning("Request did not match any route in MVC.");
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouterMiddleware.cs]: This should be verbose not warning, and routing isn't an MVC feature.

File: src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs [hunk 4]
```diff
@@ -99,84 +99,59 @@ public string RouteTemplate
...
-                    // If we got back a null value set, that means the URI did not match
-                    return;
-                }
+            if (values == null)
+            {
+                // If we got back a null value set, that means the URI did not match
 
-                var oldRouteData = context.RouteData;
+                _logger.LogVerbose(
+                        $"Request did not match the route with name '{Name}' and template '{RouteTemplate}'.");
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs]: this seems good.

File: src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs [hunk 5]
```diff
@@ -99,84 +99,59 @@ public string RouteTemplate
...
 
-                var newRouteData = new RouteData(oldRouteData);
-                MergeValues(newRouteData.DataTokens, _dataTokens);
-                newRouteData.Routers.Add(_target);
-                MergeValues(newRouteData.Values, values);
+                return;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs]: Should we log a successful match?

File: src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs [hunk 6]
```diff
@@ -99,84 +99,59 @@ public string RouteTemplate
...
+                this,
+                RouteDirection.IncomingRequest,
+                _constraintLogger))
+            {
+                return;
+            }
 
-                    await _target.RouteAsync(context);
+            _logger.LogInformation(
+                $"Request successfully matched the route with name '{Name}' and template '{RouteTemplate}'.");
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs]: This seems good.

Comment (rynowak) [inline src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs]: Should we log the route values? Probably safe to leave it out for now and see if it's necessary later.

**Comment by rynowak** [comment]
Overall seems on the right track :watch:

File: src/Microsoft.AspNet.Routing/RouterMiddleware.cs (hunk 3 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Routing/RouterMiddleware.cs]: Good point :+1:

File: src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs (hunk 6 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs]: :+1: I think we should, for example in the case where route matching succeeds but action selection fails, route values would be useful in diagnosis.

File: src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs [hunk 7]
```diff
@@ -24,35 +22,30 @@ public static class RouteConstraintMatcher
                 return true;
             }
 
+            logger.LogVerbose("Verifying route constraints");
+            if(constraints != null && constraints.Count == 0)
+            {
+                logger.LogVerbose("No route constraints found.");
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs]: None of this should be here

File: src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs [hunk 8]
```diff
@@ -24,35 +22,30 @@ public static class RouteConstraintMatcher
...
+
             foreach (var kvp in constraints)
             {
                 var constraint = kvp.Value;
+
+                logger.LogVerbose("Verifying constraint '{RouteConstraint}' for '{RouteKey}'", constraint, kvp.Key);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs]: This is tracing, we're not doing tracing.

File: src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs [hunk 9]
```diff
@@ -24,35 +22,30 @@ public static class RouteConstraintMatcher
...
-                        logger.WriteValues(new RouteConstraintMatcherMatchValues()
-                        {
-                            ConstraintKey = kvp.Key,
-                            Constraint = kvp.Value,
-                            Matched = false
-                        });
+                        logger.LogWarning(
+                            "Route value for '{RouteKey}' did not match the constraint '{RouteConstraint}'.",
+                            kvp.Key,
+                            kvp.Value);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs]: Can we include the value as well?

Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs]: This is not a warning, it's not really a mistake or a programming error, it should be verbose.

File: src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs [hunk 10]
```diff
@@ -1,15 +1,14 @@
...
-using System.Threading.Tasks;
 using Microsoft.AspNet.Http;
 using Microsoft.AspNet.Routing.Logging;
-using Microsoft.AspNet.Routing.Logging.Internal;
 using Microsoft.Framework.DependencyInjection;
 using Microsoft.Framework.Internal;
 using Microsoft.Framework.Logging;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs]: Sort

File: src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs [hunk 11]
```diff
@@ -1,12 +1,10 @@
 // Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-using System.Collections.Generic;
 using Microsoft.AspNet.Http;
-using Microsoft.AspNet.Routing.Logging;
-using Microsoft.AspNet.Routing.Logging.Internal;
 using Microsoft.Framework.Internal;
 using Microsoft.Framework.Logging;
+using System.Collections.Generic;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs]: Sort

File: src/Microsoft.AspNet.Routing/Logging/RouteValuesLogValues.cs [hunk 12]
```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.Framework.Logging;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNet.Routing.Logging
+{
+    public class RouteValuesLogValues : ILogValues
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/Logging/RouteValuesLogValues.cs]: I get why you're doing this, but it's a deviation from the pattern. Let's get buy-in to this being necessary and good.

File: src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs [hunk 13]
```diff
@@ -99,84 +98,71 @@ public string RouteTemplate
...
+                return;
+            }
 
-                    return;
-                }
+            _logger.LogVerbose(
+                    "Request successfully matched the template of the route with name '{RouteName}' and template" +
+                    " '{RouteTemplate}'.",
+                    Name,
+                    RouteTemplate);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs]: I think we should make this be 'information'

Comment (rynowak) [inline src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs]: I think we can skip this.

**Comment by rynowak** [comment]
:watch:

**Comment by davidfowl** [comment]
Well no the builds are failing

**Comment by rynowak** [comment]
@davidfowl stop commenting on PRs and come to the meeting you set up :laughing:

File: src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs [hunk 14]
```diff
@@ -29,30 +27,21 @@ public static class RouteConstraintMatcher
...
-                        {
-                            ConstraintKey = kvp.Key,
-                            Constraint = kvp.Value,
-                            Matched = false
-                        });
+                        object routeValue;
+                        routeValues.TryGetValue(kvp.Key, out routeValue);
+
+                        logger.LogVerbose(
+                            "Route value '{RouteValue}' of key '{RouteKey}' did not match " +
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteConstraintMatcher.cs]: of key -> with key

File: test/Microsoft.AspNet.Routing.Tests/Template/TemplateRouteTest.cs [hunk 15]
```diff
@@ -65,132 +82,62 @@ public void CreateTemplate_InlineConstraint_Regex_Malformed()
...
-            Assert.Equal(template, values.Template);
-            Assert.NotNull(values.DefaultValues);
-            Assert.NotNull(values.ProducedValues);
-            Assert.Equal(true, values.MatchedTemplate);
-            Assert.Equal(false, values.MatchedConstraints);
-            Assert.Equal(false, values.Matched);
-            Assert.Equal(context.IsHandled, values.Handled);
+            Assert.Empty(sink.Scopes);
+            Assert.Single(sink.Writes);
+            Assert.Equal(expectedMessage, sink.Writes[0].State?.ToString());
```
Comment (rynowak) [inline test/Microsoft.AspNet.Routing.Tests/Template/TemplateRouteTest.cs]: Let's also verify the key-values here.

**Comment by rynowak** [comment]
:watch: for test improvements

**Comment by rynowak** [comment]
:shipit: 

Think about commonizing the code that you're using to compare dictionaries, this is something that's not trivial in xunit, and you'll be doing a fair bit of it.

**Comment by kichalla** [comment]
Thanks...sure will work on creating an assert for dictionaries...

**Pull request merged**